### PR TITLE
Preload fonts in the BO

### LINKED ISSUE #26837: Flash of unstyled text while material icons load causes weird menu
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md)
- [X] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add screenshots

When the page is loaded with a clean cache, there's a flash of unstyled content for a second or two while the menu icons are loading:

Screenshot 2021-12-02 at 13 02 06 (2)

### Expected behavior

The "Flash of unstyled text" is well-documented on the web, and there several approaches to work around it. I suggest finding a way not to display the icons while the font is loading, perhaps by using `font-display: block`

### Steps to reproduce

1. Load the BO with a clean browser cache and/or a slow connection
2. On first, load, see the flashing text while the icons are being loaded

### PrestaShop version(s) where the bug happened

At least since 1.7.7, probably existed since 1.7.5 and maybe even before

### PHP version(s) where the bug happened

7.4

### If your bug is related to a module, specify its name and its version

_No response_

### PR DESCRIPTION
**Pull request by NeOMakinG**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Font loading is adding flash of content, specially on menu
| Type?             | bug fix
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Fixes #26837 (Only partially on good connection, not a full fix).
| How to test?      | Try to load the BO, check if the material icon lib is preloaded (inside your network tab)
| Possible impacts? | Font loading of the BO

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/26883)

### CODE REVIEW

**Comment by mparvazi** [comment]
@NeOMakinG 

Page load time in seconds (Increased time %)
Firefox Developer Tools 

| Preload  | Cached | Good 3G | Regular 4G / LTE | Wi-Fi |
| ------------- | ------------- | ------------- | ------------- | ------------- |
| YES  | NO  | 22.50 (+50%)  | 10.20 (+55%) | 2.42 (+50%)  |
| NO  | NO | 11.40 | 4.60 | 1.25  |
| YES  | YES  | 18.94 (+70%) | 9.31 (+75%)  | 2.16 (+75%) |
| NO  | YES | 6.26 | 2.48 | 0.75  |


### Test without speed limitation

**With preload:**



**Without preload:**

File: admin-dev/themes/default/webpack.config.js
Comment (Oksydan) [inline admin-dev/themes/default/webpack.config.js]: @NeOMakinG preload only woff2 format. You are loading every single file on page load.

**Review by Oksydan — COMMENTED** [review]
@NeOMakinG small feedback.

File: admin-dev/themes/default/webpack.config.js
Comment (NeOMakinG) [reply admin-dev/themes/default/webpack.config.js]: Damn, mb 🤣

**Comment by mparvazi** [comment]
Browser support for webfonts:

**EOT** — compatible with Internet Explorer 6+.

**WOFF** — compatible with Firefox 3.6+, Safari 5.1+, Chrome 5+, Internet Explorer 9+, Opera 11.5+, Android 4.4+, and iOS Safari 5.1+.

**WOFF2** — The next-generation of WOFF providing a compression advantage of roughly 20–30% off the font’s file size. Compatible with Firefox 3.6+, Safari 5.1+, Chrome 36+, Edge 14+, Opera 23+, Android 67+, and iOS Safari 10+ (Mac OS Sierra and up only).

Good news!
The Internet Explorer will be retired and go out of support on June 15, 2022.

**Comment by NeOMakinG** [comment]
> Browser support for webfonts:
> 
> **EOT** — compatible with Internet Explorer 6+.
> 
> **WOFF** — compatible with Firefox 3.6+, Safari 5.1+, Chrome 5+, Internet Explorer 9+, Opera 11.5+, Android 4.4+, and iOS Safari 5.1+.
> 
> **WOFF2** — The next-generation of WOFF providing a compression advantage of roughly 20–30% off the font’s file size. Compatible with Firefox 3.6+, Safari 5.1+, Chrome 36+, Edge 14+, Opera 23+, Android 67+, and iOS Safari 10+ (Mac OS Sierra and up only).
> 
> Good news! The Internet Explorer will be retired and go out of support on June 15, 2022.

The project should only support IE11, and tbh I don't give a fuck about supporting it anymore myself, the usage is way too little...

Anyway, I think it will load needed fonts but not preloaded so

**Comment by mparvazi** [comment]
@NeOMakinG 

> Anyway, I think it will load needed fonts but not preloaded so

Right!

```
<head>
    <link rel="preload" as="font" href="/somefolder/font-one.woff2">
</head>
```

This way woff2 will be preloaded by browsers that support preload, and all the fallback formats will load as they normally do.

```
@font-face {
    font-family: FontOne;
    src: url(../somefolder/font-one.eot);
    src: url(../somefolder/font-one.eot?#iefix) format('embedded-opentype'),
    url(../somefolder/font-one.woff2) format('woff2'), //Will be preloaded
    url(../somefolder/font-one.woff) format('woff'),
    url(../somefolder/font-one.ttf)  format('truetype'),
    url(../somefolder/font-one.svg#svgFontName) format('svg'); 
}
```

**Comment by Oksydan** [comment]
@mparvazi 
It is working this way but it automated via webpack plugin.
File will be generated automatically on webpack build and  included inside BO head.

**Comment by mparvazi** [comment]
@NeOMakinG 

theme.css

With PR:

```
@font-face {
	font-family: "Material Icons";
	font-style: normal;
	font-weight: 400;
	src: url(2a61b00fcc9b1afbf99be145777b1b72.eot);
	src: local("Material Icons"), local("MaterialIcons-Regular"), 
		url(703cf8f274fbb265d49c6262825780e1.woff2) format("woff2"), 
		url(84a37de85c17f186652a179b1145392f.woff) format("woff"), 
		url(51cf1d641ae503ece727e7d84b182524.ttf) format("truetype")
}
```

Without PR:

```
@font-face {
	font-family: "Material Icons";
	font-style: normal;
	font-weight: 400;
	font-display: block;
	src: url(2a61b00fcc9b1afbf99be145777b1b72.eot);
	src: local("Material Icons"), local("MaterialIcons-Regular"), 
		url(703cf8f274fbb265d49c6262825780e1.woff2) format("woff2"), 
		url(84a37de85c17f186652a179b1145392f.woff) format("woff"), 
		url(51cf1d641ae503ece727e7d84b182524.ttf) format("truetype")
}
```

What does it means?
EOT should be preloaded?

**Comment by mparvazi** [comment]
| Preload  | Cached | Good 3G | Regular 4G / LTE | Wi-Fi |
| ------------- | ------------- | ------------- | ------------- | ------------- |
| YES  | NO  | 12.55 (+10%)  | 5.65 (+20%) | 1.97 (+40%)  |
| NO  | NO | 11.40 | 4.60 | 1.25  |
| YES  | YES  | 8.95 (+40%) | 3.95 (+40%)  | 1.54 (+50%) |
| NO  | YES | 6.26 | 2.48 | 0.75  |

**Comment by NeOMakinG** [comment]
> Preload	Cached	Good 3G	Regular 4G / LTE	Wi-Fi
> YES	NO	12.55 (+10%)	5.65 (+20%)	1.97 (+40%)
> NO	NO	11.40	4.60	1.25
> YES	YES	8.95 (+40%)	3.95 (+40%)	1.54 (+50%)
> NO	YES	6.26	2.48	0.75

I guess it's acceptable, no?

Anyway, the CSS fix would be fine for the menu, but not for any others single icons so..

**Comment by mparvazi** [comment]
About one second is added to the page load time to preload fonts.
Fixing by CSS for the menu is easy, but needs huge changes in templates for other icons, so this solution is much better!

My problem:
Why browser in cache enabled state, checks preload fonts for modification?



Why doesn't check cache before?
It sends request to server to check if font is modified, and if not modified will use cache.

I think it's maybe related to Apache handlers for cache-control.

**Comment by NeOMakinG** [comment]
> About one second is added to the page load time to preload fonts. Fixing by CSS for the menu is easy, but needs huge changes in templates for other icons, so this solution is much better!
> 
> My problem: Why browser in cache enabled state, checks preload fonts for modification?
> 
> 
> 
> Why doesn't check cache before?
> 
> I think it's maybe related to Apache handlers for cache-control.
> 
> 

Sorry but I won't change the icon value of alone icons (dashboard, in-card icons, grids), because it's not good at all for the accessibility.

Only icons that can be changed are icons with a text (menu icons for example), because they've a label which describe them so it shouldn't be a problem for the accessibility.

We can do this on the menu, that could be a partial fix, but I won't do your fix on icons if they are not described by a text => screen readers wouldn't be able to describe the icons...

**Comment by mparvazi** [comment]
@NeOMakinG 

This PR works fine!
I think there should be a way to prevent browser for checking fonts modification.
In this case, no extra time is added to page load in cached state.

**Comment by mparvazi** [comment]
@NeOMakinG 

Improvement:

Just `703cf8f274fbb265d49c6262825780e1.woff2` is related to material icons.
Other fonts shouldn't be included in `preload.tpl`

**FontAwesome**
97493d3f11c0a3bd5cbd959f5d19b699.woff2

**Open Sans Font Family**
55835483c304eaa8477fea2c36abba17.woff2
383eba0e55ed778006d76428812d343c.woff2
08952b029e4decbc8ef9fb553cae8cea.woff2
3343e54368719e3786f78a1b22839455.woff2
3343e54368719e3786f78a1b22839455.woff2
3326e4d74d3924ee1c882c29f5b571c0.woff2
5aaceea2d60ddb477c6aafc825eece3d.woff2

**Ubuntu Condensed**
eb24af6668c633feab6bf8f989296e73.woff2

```
<link
      rel="preload"
      href="{"`$admin_dir`"}703cf8f274fbb265d49c6262825780e1.woff2"
      as="font"
      crossorigin
    ><link
      rel="preload"
      href="{"`$admin_dir`"}97493d3f11c0a3bd5cbd959f5d19b699.woff2"
      as="font"
      crossorigin
    ><link
      rel="preload"
      href="{"`$admin_dir`"}55835483c304eaa8477fea2c36abba17.woff2"
      as="font"
      crossorigin
    ><link
      rel="preload"
      href="{"`$admin_dir`"}383eba0e55ed778006d76428812d343c.woff2"
      as="font"
      crossorigin
    ><link
      rel="preload"
      href="{"`$admin_dir`"}08952b029e4decbc8ef9fb553cae8cea.woff2"
      as="font"
      crossorigin
    ><link
      rel="preload"
      href="{"`$admin_dir`"}3343e54368719e3786f78a1b22839455.woff2"
      as="font"
      crossorigin
    ><link
      rel="preload"
      href="{"`$admin_dir`"}3326e4d74d3924ee1c882c29f5b571c0.woff2"
      as="font"
      crossorigin
    ><link
      rel="preload"
      href="{"`$admin_dir`"}5aaceea2d60ddb477c6aafc825eece3d.woff2"
      as="font"
      crossorigin
    ><link
      rel="preload"
      href="{"`$admin_dir`"}eb24af6668c633feab6bf8f989296e73.woff2"
      as="font"
      crossorigin
    >
```

The preload will be limited to one font and the speed result is great!

**Comment by mparvazi** [comment]
| Preload  | Cached | Good 3G | Regular 4G / LTE | Wi-Fi |
| ------------- | ------------- | ------------- | ------------- | ------------- |
| YES (All fonts)  | NO  | 12.55 (+10%)  | 5.65 (+20%) | 1.97 (+40%)  |
| YES (Material)  | NO  | 11.55 (+2%)  | 4.85 (+6%) | 1.71 (+25%)  |
| NO  | NO | 11.40 | 4.60 | 1.25  |
| YES (All fonts)  | YES  | 8.95 (+40%) | 3.95 (+40%)  | 1.54 (+50%) |
| YES (Material)  | YES  | 6.97 (+20%) | 3.05 (+20%)  | 1.31 (+40%) |
| NO  | YES | 6.26 | 2.48 | 0.75  |

**Comment by mparvazi** [comment]
@NeOMakinG 

It's your magic how to preload just `MaterialIcons-Regular.woff2`

**Comment by NeOMakinG** [comment]
> @NeOMakinG
> 
> It's your magic how to preload just `MaterialIcons-Regular.woff2`

Gonna check out how to do this

**Comment by NeOMakinG** [comment]
Managed to preload only the material-icons font but I've a problem:

- FontPreloadPlugin doesn't permit us to flag the file we want to preload, we need to extend it, but latest release was in may and the repository is not that active, it could take a lot of time to get it merged
- VueJs maintained PreloadPlugin doesn't permit to append the links into something else than the head of an HTML Tag and I think it wouldn't be compatible with our project

We've 2-3 solutions:
- Add the file as static inside the project
- Fork FontPreloadPlugin and maintain it as an NPM module
- Contribute to the project (but can be really slow and refused)

I wan't you advice @PrestaShop/prestashop-maintainers

**Comment by mparvazi** [comment]
Emulator: Firefox Developer Tools
Test page: Dashboard
Debug mode: Off

**Total time needed to load all requests**
| Preload  | Cache | Good 3G | Regular 4G / LTE | Wi-Fi |
| ------------- | ------------- | ------------- | ------------- | ------------- |
| YES  | Empty  | 11.74 | 5.04 | 2.21 |
| NO  | Empty | 11.58 | 4.98 | 2.17 |
| YES  | Primed  | 7.58 | 3.82 | 1.76 |
| NO  | Primed | 7.30 | 3.70 | 1.72 |


**Time spent for preloading (Empty cache)**
| Good 3G | Regular 4G / LTE | Wi-Fi |
| ------------- | ------------- | ------------- |
| 80ms | 50ms | 33ms |

**Comment by NeOMakinG** [comment]
> Emulator: Firefox Developer Tools Test page: Dashboard Debug mode: Off
> 
> **Total time needed to load all requests**
> 
> Preload	Cache	Good 3G	Regular 4G / LTE	Wi-Fi
> YES	Empty	11.74	5.04	2.21
> NO	Empty	11.58	4.98	2.17
> YES	Primed	7.58	3.82	1.76
> NO	Primed	7.30	3.70	1.72
> **Time spent for preloading (Empty cache)**
> 
> Good 3G	Regular 4G / LTE	Wi-Fi
> 80ms	50ms	33ms

Dude, is it longer than loading all fonts? There is a problem somewhere

**Comment by mparvazi** [comment]
@NeOMakinG 

I think there is no problem, the page load is fast.

For testing I had to refresh the page every time to get the speed results (I wanted to test the hardest possible case for speed). 
The values for page load time in table can be consider as reference for compare, in reality it's much faster.
Don't compare the values for time between tables that I sent.
The emulator shows different values each time, so just focus on differences between values inside each table.

Just 30ms will be added to page load in empty cache. 
So tell me who cares about adding this time only once after installing PS.
Then just after first load, it will be zero!

The way you just added the material icons font to the preload was great!

**Comment by NeOMakinG** [comment]
> @NeOMakinG
> 
> I think there is no problem, the page load is fast.
> 
> For testing I had to refresh the page every time to get the speed results (I wanted to test the hardest possible case for speed). The values for page load time in table can be consider as reference for compare, in reality it's much faster. Don't compare the values for time between tables that I sent. The emulator shows different values each time, so just focus on differences between values inside each table.
> 
> Just 30ms will be added to page load in empty cache. So tell me who cares about adding this time only once after installing PS. Then just after first load, it will be zero!
> 
> The way you just added the material icons font to the preload was great!

Ok, discussing with the maintainer of the webpack plugin so I can contribute and push this further

https://github.com/sn-satyendra/webpack-font-preload-plugin/issues/26

**Comment by NeOMakinG** [comment]
We should be ready now :)

**Comment by mparvazi** [comment]
@NeOMakinG 

lemme check :))

**Comment by mparvazi** [comment]
### Empty cache

Screenshot from 2021-12-13 15-35-23

**Time spent for preload (Empty cache):**
| Good 3G | Regular 4G / LTE | Wi-Fi |
| ------------- | ------------- | ------------- |
| 77 ms | 48 ms | 27 ms |


### Primed cache

Screenshot from 2021-12-13 15-36-46

**Time spent for preload (Primed cache):**
| Good 3G | Regular 4G / LTE | Wi-Fi |
| ------------- | ------------- | ------------- |
| 0 ms | 0 ms | 0 ms |

**Review by Progi1984 — DISMISSED** [review]

File: admin-dev/themes/default/webpack.config.js
Comment (PierreRambaud) [inline admin-dev/themes/default/webpack.config.js]:
```suggestion
        test: /\.(jpg|png|woff2?|eot|otf|ttf|svg|gif)$/,
```

File: admin-dev/themes/default/webpack.config.js
Comment (PierreRambaud) [inline admin-dev/themes/default/webpack.config.js]:
```suggestion
        exclude: /MaterialIcons-Regular\.(woff2?|ttf)$/,
```

File: admin-dev/themes/default/webpack.config.js
Comment (PierreRambaud) [inline admin-dev/themes/default/webpack.config.js]:
```suggestion
        test: /MaterialIcons-Regular\.(woff2?|ttf)$/,
```

File: admin-dev/themes/new-theme/.webpack/common.js
Comment (PierreRambaud) [inline admin-dev/themes/new-theme/.webpack/common.js]:
```suggestion
        exclude: /MaterialIcons-Regular\.(woff2?|ttf)$/,
```

File: admin-dev/themes/new-theme/.webpack/common.js
Comment (PierreRambaud) [inline admin-dev/themes/new-theme/.webpack/common.js]:
```suggestion
        test: /MaterialIcons-Regular\.(woff2?|ttf)$/,
```

File: admin-dev/themes/new-theme/.webpack/common.js
Comment (PierreRambaud) [inline admin-dev/themes/new-theme/.webpack/common.js:389]:
```suggestion
      templateContent: '{{{preloadLinks}}}',
```

Any reason for the triple `{`?

Comment (NeOMakinG) [reply admin-dev/themes/new-theme/.webpack/common.js:389]: We could change it to anything else, but {{{}}} doesn't exist anywhere else, I find it pretty cool to see something else than any known syntax as a "template string"

**Review by PierreRambaud — DISMISSED** [review]

**Review by Progi1984 — DISMISSED** [review]

**Comment by HanaRebaiQA** [comment]
Hello @NeOMakinG 

I check this PR and i have this issue related to `webpack-cli` under `themes` folder during the make install.
image

Could you check it please and feedback!

Thanks!

**Comment by HanaRebaiQA** [comment]
Hello @mparvazi 

Thanks for your information.
Im used theses versions and not worked fine : 
node.js : v10.19.0
npm : 7.24.2
php : 7.3
image

Please check and waiting for feedback!

Thanks!

**Comment by mparvazi** [comment]
Hello @HanaRebaiQA 

node: v10.24.1
npm: v6.14.12
php : 8.0

Screenshot from 2021-12-16 11-46-44

https://github.com/nvm-sh/nvm

**Comment by NeOMakinG** [comment]
@HanaRebaiQA please update to node 14, it will work both on 1.7.8 and develop

**Comment by mparvazi** [comment]
@NeOMakinG :thinking:

**Comment by mparvazi** [comment]
Hello @HanaRebaiQA 
Could you please check again?
There is no problem in PR.

**Comment by NeOMakinG** [comment]
@HanaRebaiQA 

We really want to merge this PR, it will greatly improve the loading time of initial requests in the BO, even if it doesn't full fix the initial bug, it's a great improvement that we would love to have.

**Review by PierreRambaud — APPROVED** [review]

**Review by PululuK — APPROVED** [review]

**Comment by HanaRebaiQA** [comment]
Hello @NeOMakinG 

I checked again this PR. As you mentioned above, the issue is partially fixed.
https://watch.screencastify.com/v/O7JgfurjHEqNRAxP5ldD

So, with a good connection, the proload fonts is OK.
https://watch.screencastify.com/v/aZV5vN26QJoiCrWRlclS

So, it's QA :heavy_check_mark: 

Thanks!

**Comment by Progi1984** [comment]
Thanks @NeOMakinG & @HanaRebaiQA

**Pull request merged**
